src/blocks/weather/open_weather_map.rs

```diff
@@ -79,24 +94,233 @@ struct ApiWeather {
     description: String,
 }
 
-#[async_trait]
-impl WeatherProvider for Service<'_> {
-    async fn get_weather(&self, autolocated: Option<Coordinates>) -> Result<WeatherResult> {
-        let api_key = self.config.api_key.as_ref().or_error(|| {
-            format!("missing key 'service.api_key' and environment variable {API_KEY_ENV}",)
-        })?;
+#[derive(Deserialize, Debug)]
+struct City {
+    //This should be an u32, but the source data has a few errors
+    id: f32,
+    coord: CityCoord,
+}
+#[derive(Deserialize, Debug)]
+struct CityCoord {
+    lat: f64,
+    lon: f64,
+}
+
+#[derive(Eq, Hash, PartialEq, Debug)]
+enum LocationKeys {
+    CityId(String),
+    Place(String),
+    Zip(String),
+}
 
-        let location_query = autolocated
+static LOCATION_CACHE: Lazy<Mutex<HashMap<LocationKeys, String>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
```

Comment:
Why can't we just store the cached location string in the `Service` struct?

Reply:
I tried that initially, but could not figure out how to fix the error caused when you make the service mutable. The line above the retry becomes a FnMut, and complains: `Captured variable cannot escape FnMut closure body`

Reply:
Oh, yeah, that's an annoying error. Can we construct the location string in `Service::new()` then? It can take `autolocate: bool` and do the necessary stuff if it is `false`.